File: packages/data/lib/src/dtos/address_dto.dart
import 'package:banan_domain/banan_domain.dart';

class AddressDto {
  const AddressDto({
    required this.id,
    required this.label,
    required this.recipient,
    required this.phone,
    required this.line1,
    required this.city,
    this.line2,
    this.district,
    this.wardCode,
    this.postalCode,
    this.isDefault = false,
  });

  factory AddressDto.fromJson(Map<String, dynamic> json) {
    return AddressDto(
      id: json['id'] as String,
      label: json['label'] as String? ?? 'Delivery',
      recipient: json['recipient'] as String,
      phone: json['phone'] as String,
      line1: json['line1'] as String,
      line2: json['line2'] as String?,
      city: json['city'] as String,
      district: json['district'] as String?,
      wardCode: json['wardCode'] as String?,
      postalCode: json['postalCode'] as String?,
      isDefault: json['isDefault'] as bool? ?? false,
    );
  }

  final String id;
  final String label;
  final String recipient;
  final String phone;
  final String line1;
  final String? line2;
  final String city;
  final String? district;
  final String? wardCode;
  final String? postalCode;
  final bool isDefault;

  /// Build a domain Address. Caller may pass a [wardName] looked up from
  /// the ward catalog so list views render the new ward label instead of
  /// the legacy district.
  Address toDomain({String? wardName}) => Address(
        id: id,
        label: label,
        recipient: recipient,
        phone: phone,
        line1: line1,
        line2: line2,
        city: city,
        district: district,
        wardCode: wardCode,
        wardName: wardName,
        postalCode: postalCode,
        isDefault: isDefault,
      );
}

File: backend/src/addresses/dto/address.dto.ts
import {
  IsBoolean,
  IsOptional,
  IsString,
  MaxLength,
  MinLength,
} from 'class-validator';

export class CreateAddressDto {
  @IsString()
  @MinLength(1)
  @MaxLength(40)
  label!: string;

  @IsString()
  @MinLength(1)
  @MaxLength(120)
  recipient!: string;

  @IsString()
  @MinLength(1)
  @MaxLength(20)
  phone!: string;

  @IsString()
  @MinLength(1)
  @MaxLength(160)
  line1!: string;

  @IsOptional()
  @IsString()
  @MaxLength(80)
  line2?: string;

  @IsString()
  @MinLength(1)
  @MaxLength(80)
  city!: string;

  @IsOptional()
  @IsString()
  @MaxLength(80)
  district?: string;

  /// HCMC post-2025 ward code. Required for delivery in HCMC so we can
  /// compute the distance-based delivery surcharge.
  @IsOptional()
  @IsString()
  @MaxLength(60)
  wardCode?: string;

  @IsOptional()
  @IsString()
  @MaxLength(20)
  postalCode?: string;

  @IsOptional()
  @IsBoolean()
  isDefault?: boolean;
}

export class UpdateAddressDto {
  @IsOptional()
  @IsString()
  @MinLength(1)
  @MaxLength(40)
  label?: string;

  @IsOptional()
  @IsString()
  @MinLength(1)
  @MaxLength(120)
  recipient?: string;

  @IsOptional()
  @IsString()
  @MinLength(1)
  @MaxLength(20)
  phone?: string;

  @IsOptional()
  @IsString()
  @MinLength(1)
  @MaxLength(160)
  line1?: string;

  @IsOptional()
  @IsString()
  @MaxLength(80)
  line2?: string;

  @IsOptional()
  @IsString()
  @MinLength(1)
  @MaxLength(80)
  city?: string;

  @IsOptional()
  @IsString()
  @MaxLength(80)
  district?: string;

  @IsOptional()
  @IsString()
  @MaxLength(60)
  wardCode?: string;

  @IsOptional()
  @IsString()
  @MaxLength(20)
  postalCode?: string;

  @IsOptional()
  @IsBoolean()
  isDefault?: boolean;
}
